Add UUID To RequestContext (#2001)

yanghun0070 opened the pull request:
Motivation:
- It would be useful, and probably simple, to assign a UUID to every request to add to the log to be able to associate requests and responses. 

Modifications:
- Add  `UUID` to RequestContext.

Result:
-  assign a different UUID to every requests and responses context.

Commit:
Add UUID To RequestContext

Commit:
Add UUID To RequestContext

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/line/armeria?pullRequest=2174) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/line/armeria?pullRequest=2174) before we can accept your contribution.<br/>**9** out of **10** committers have signed the CLA.<br/><br/>:white_check_mark: ikhoon<br/>:white_check_mark: yanghun0070<br/>:white_check_mark: trustin<br/>:white_check_mark: minwoox<br/>:white_check_mark: anuraaga<br/>:white_check_mark: heowc<br/>:white_check_mark: taejs<br/>:white_check_mark: matsumana<br/>:white_check_mark: jyblue<br/>:x: jeehun.yang<br/><hr/>**jeehun.yang** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/line/armeria?pullRequest=2174) it.</sub>

trustin commented:
@yanghun0070 Could you please rebase your work?

yanghun0070 committed:
Merge branch 'master' into feature/AddUUIDFromRequest

yanghun0070 commented:
@trustin I already rebase my work.

trustin commented on brave/src/test/java/com/linecorp/armeria/client/brave/BraveClientIntegrationTest.java:
```diff
@@ -163,7 +164,12 @@ protected void getAsync(HttpClient client, String pathIncludingQuery) throws Exc
+        @Override
+        public RequestContext newDerivedContext() {
+            throw new UnsupportedOperationException();
         }
```
You don't need to override this method.

trustin commented on core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java:
```diff
@@ -433,6 +437,13 @@ protected final String path() {
+     */
+    protected final UUID uuid() {
+        return uuid;
+    }
```
- Could you add a public setter of the `uuid` as well?
- We could start with `null` `UUID` and then call `UUID.randomUUID()` only when a user did not call `builder.uuid(UUID)` (i.e. `this.uuid == null`).

trustin commented on core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java:
```diff
@@ -54,6 +55,7 @@
     private volatile HttpRequest req;
     @Nullable
     private volatile RpcRequest rpcReq;
+    protected UUID uuid;
```
- `private final` unless there is a good reason to make it mutable and protected.
- I guess we need to move this field right before `method`, because it is an essential non-null per-request field.

trustin commented on core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java:
```diff
@@ -67,18 +69,20 @@
     protected NonWrappingRequestContext(
             MeterRegistry meterRegistry, SessionProtocol sessionProtocol,
-            HttpMethod method, String path, @Nullable String query,
+            HttpMethod method, String path, UUID uuid, @Nullable String query,
```
How about moving `uuid` right before `method`? Please make the same change for all `RequestContext` constructors you modified.

trustin commented on core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java:
```diff
@@ -67,18 +69,20 @@
         this.sessionProtocol = requireNonNull(sessionProtocol, "sessionProtocol");
         this.method = requireNonNull(method, "method");
         this.path = requireNonNull(path, "path");
+        this.uuid = requireNonNull(uuid, "uuid");
```
Ditto - Please move this assignment right before `method` assignment. Same for all similar changes in this PR.

trustin commented on core/src/main/java/com/linecorp/armeria/common/RequestContext.java:
```diff
@@ -191,6 +192,11 @@
+    /**
+     * Returns the {@link UUID} that contains the information about the current {@link Request}.
+     */
+    UUID uuid();
```
Could you move this method right before `method()`?

trustin commented on core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java:
```diff
@@ -178,6 +182,11 @@ public final String query() {
+    @Override
+    public final UUID uuid() {
+        return uuid;
+    }
```
Could you move this method right before `method()`?

trustin commented on core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java:
```diff
@@ -159,7 +160,7 @@ private DefaultServiceRequestContext(
 
         super(meterRegistry, sessionProtocol,
-              requireNonNull(routingContext, "routingContext").method(), routingContext.path(),
+              requireNonNull(routingContext, "routingContext").method(), routingContext.path(), UUID.randomUUID(),
```
Please make `Default{Client,Service}RequestContext` have symmetric and consistent API, i.e. Let a user specify an UUID as a parameter just like you did in `DefaultClientRequestContext`.

trustin reviewed (changes requested):
When creating a derived context, a user might want or *not* want to assign a new UUID to the derived context. How about adding `UUID uuid` as an additional parameter of `newDerivedContext(HttpRequest, RpcRequest)` and `newDerivedContext(HttpRequest, RpcRequest, Endpoint)`?

I'm somewhat unsure about what the behavior of `newDerivedContext()` is supposed to be. Should we:

- Always assign a new random UUID,
- Add `UUID` as a parameter there as well or
- Just remove `newDerivedContext()` to get rid of ambiguity..?

Let me know what you think. @minwoox @ikhoon @anuraaga

trustin commented:
And perhaps we could add `RequestLog.uuid()` getter as well, which is a simple shortcut to `RequestLog.context().uuid()`.

minwoox commented:
> Just remove newDerivedContext() to get rid of ambiguity..?

+1 for this.
IIRC, the `newDerivedContext` is requested by one of our customers who was not using `RetryingClient`. He implemented his own retrying client because at that time `Retryingclient` had a bug that uses only one `requestLog` no matter how many retry is made so he could not log the causes and responses successfully.
But right now, we fixed it and it works well. I don't think we need it anymore.

ikhoon committed:
Drop `armeria-zipkin` artifact from setup.html (#2175)

Motivation:
`armeira-zipkin` has been removed since
[0.94.0 release](https://github.com/line/armeria/releases/tag/armeria-0.94.0).

Modification:
*  Replace `armeria-zikin` with  `armeria-brave` in documentation

Result:
A user can set up `armeria-brave` using
[`Setting up a project`](https://line.github.io/armeria/setup.html#setting-up-a-project) guide.

trustin committed:
Remove the workaround for gradle/gradle#1365 (#2178)

.. because it's been fixed in Gradle 5.6

heowc committed:
Provide a way to configure a CacheSpec for HttpFileService (#2142)

Motivation:

A user cannot set a way to configure a CacheSpec for HttpFileService.

Modifications:

- Add a new flag `com.linecorp.armeria.fileServiceCache=<spec>`
- Add `entryCacheSpec` for `HttpFileServiceBuilder`
- Delegate `maxCacheEntries` to `entryCacheSpec`

Result:

- Fixes #916

yanghun0070 committed:
Fix consistence to moving uuid right before method
Fix DefaultServiceRequestContext have symmetric DefaultClientRequestContext

yanghun0070 committed:
Fix consistence to moving uuid right before method
Fix DefaultServiceRequestContext have symmetric DefaultClientRequestContext

minwoox committed:
Fix a bug where all Checkstyle rules are suppressed (#2183)

anuraaga committed:
Migrate debug page to hooks. (#2139)

Also fixes long-time pet peeve by prettifying gRPC unframed JSON.

Fixes #1949

yanghun0070 committed:
Fix UUID to this method

yanghun0070 committed:
Fix UUID to this method

minwoox commented on benchmarks/src/jmh/java/com/linecorp/armeria/server/RoutersBenchmark.java:
```diff
@@ -62,7 +63,7 @@
     @Benchmark
     public Routed<ServiceConfig> exactMatch() {
-        final RoutingContext ctx = DefaultRoutingContext.of(HOST, "localhost", METHOD1_HEADERS.path(),
+        final RoutingContext ctx = DefaultRoutingContext.of(HOST, "localhost", UUID.randomUUID(), METHOD1_HEADERS.path(),
```
Seems like this line is longer than 112 which we constraint using checkstyle.
Could you rebase master and do `./gradlwe checkstyle` so that you can notice the violation?

minwoox replied:
How about making a field for `uuid` and using it here?

trustin replied:
+1

minwoox commented on brave/src/test/java/com/linecorp/armeria/client/brave/BraveClientIntegrationTest.java:
```diff
@@ -163,7 +164,7 @@ protected void getAsync(HttpClient client, String pathIncludingQuery) throws Exc
         DummyRequestContext() {
             super(NoopMeterRegistry.get(), SessionProtocol.HTTP,
-                  HttpMethod.GET, "/", null, HttpRequest.streaming(HttpMethod.GET, "/"), null);
+                  UUID.randomUUID(), HttpMethod.GET, "/",  null, HttpRequest.streaming(HttpMethod.GET, "/"), null);
```
Could you remove extra space after `"/",`?

trustin replied:
This is not fixed yet.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:
```diff
@@ -97,15 +98,16 @@
      *
      * @param eventLoop the {@link EventLoop} associated with this context
      * @param sessionProtocol the {@link SessionProtocol} of the invocation
+     * @param uuid the {@link UUID} that contains the information about the current {@link HttpRequest}.
```
`uuid the {@link UUID} that represents the current {@link Request} and {@link Response}`

trustin replied:
`... represents the unique identifier of the current {@link Request} and {@link Response} pair.`

minwoox commented on core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java:
```diff
@@ -260,7 +263,6 @@ private DefaultClientRequestContext(DefaultClientRequestContext ctx,
         if (ctx.rpcRequest() != null) {
             requireNonNull(rpcReq, "rpcReq");
         }
-
```
revert?

minwoox commented on core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java:
```diff
@@ -433,6 +436,24 @@ protected final String path() {
+    protected final UUID uuid() {
+        if(uuid == null) {
+            uuid = UUID.randomUUID();
+        }
```
We don't need this because it's always not null?

trustin replied:
Correct. `uuid` must start with `null`.

minwoox commented on core/src/main/java/com/linecorp/armeria/common/logging/RequestLog.java:
```diff
@@ -167,6 +168,13 @@ default void ensureAvailability(Iterable<RequestLogAvailability> properties) {
     RequestContext context();
 
+    /**
+     * Returns the method of the {@link Request}. This method is a shortcut to {@code context().uuid()}.
```
the method -> the uuid

trustin replied:
the method -> the UUID

minwoox commented on core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java:
```diff
@@ -159,6 +160,7 @@ private DefaultServiceRequestContext(
 
         super(meterRegistry, sessionProtocol,
+              requireNonNull(routingContext, "routingContext").uuid(),
               requireNonNull(routingContext, "routingContext").method(), routingContext.path(),
```
`requireNonNull(routingContext, "routingContext").method()` -> `requireNonNull(routingContext).method()` because we don't want to check it twice?

minwoox commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -321,9 +322,8 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
-
         final RoutingContext routingCtx =
-                DefaultRoutingContext.of(host, hostname, pathAndQuery.path(), pathAndQuery.query(),
+                DefaultRoutingContext.of(host, hostname, UUID.randomUUID(), pathAndQuery.path(), pathAndQuery.query(),
```
I think we don't need this `uuid` for `RoutingContext`. If we need that we have to use the one from the `requestLog`. WDYT? /cc @trustin, @ikhoon

trustin replied:
Agreed.

minwoox reviewed (commented):
I think we need to rebase and run the check style task 😄

minwoox commented:
From #2001, we need to leave uuid in `toStringRequestOnly` and `toStringResponseOnly`. Could you add it please?

trustin commented on core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java:
```diff
@@ -98,7 +100,7 @@ protected AbstractRequestContextBuilder(boolean server, HttpRequest req) {
         rpcReq = null;
         sessionProtocol = SessionProtocol.H2C;
-
+        uuid = UUID.randomUUID();
```
Could you begin `uuid` with `null`, and call `UUID.randomUUID()` only when a user did not set the UUID of the context?

trustin commented on core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java:
```diff
@@ -433,6 +436,24 @@ protected final String path() {
+     * Returns the {@link UUID} of the request.
+     */
+    protected final UUID uuid() {
+        if(uuid == null) {
```
Missing space between `if` and `(`

trustin commented on core/src/main/java/com/linecorp/armeria/common/NonWrappingRequestContext.java:
```diff
@@ -67,18 +69,20 @@
         this.sessionProtocol = requireNonNull(sessionProtocol, "sessionProtocol");
         this.method = requireNonNull(method, "method");
         this.path = requireNonNull(path, "path");
+        this.uuid = requireNonNull(uuid, "uuid");
```
This assignment should occur before the assignment of `method`, so that the field initialization takes place in order.

trustin commented on core/src/main/java/com/linecorp/armeria/common/RequestContext.java:
```diff
@@ -167,6 +168,11 @@
     SSLSession sslSession();
 
+    /**
+     * Returns the {@link UUID} that contains the information about the current {@link Request}.
```
`Returns the {@link UUID} that represents the unique identifier of ....` (See @minwoox's comment)

trustin committed:
Fix `ClassCastException` when calling `*HeadersBuilder.build()`… (#2193)

Motivation:

The following code, which calls `*HeadersBuilder.build()` twice, fails
with a `ClassCastException`.

    HttpHeadersBuilder b = HttpHeaders.builder();
    b.add("foo", "bar");
    b.build();
    b.build(); // Throws a ClassCastException

Modifications:

- Make sure the `parent` is of a desired type. Rebuild if not.

Result:

- Fixes #2190

yanghun0070 committed:
Remove UUID from RoutingContext

yanghun0070 committed:
Remove UUID from RoutingContext

taejs committed:
Add favicon to DocService (#2186)

Motivation:

To make DocService easier to identify

Modifications:

- Add a favicon to `DocService` and the official web site.

Result:

before
![image](https://user-images.githubusercontent.com/41318449/66719012-5f34b080-ee25-11e9-8778-748db2f7ee64.png)

after
![image](https://user-images.githubusercontent.com/41318449/66719008-4fb56780-ee25-11e9-83af-c2cde4dbcf3c.png)

trustin commented on benchmarks/src/jmh/java/com/linecorp/armeria/server/RoutersBenchmark.java:
```diff
@@ -17,6 +17,7 @@
 package com.linecorp.armeria.server;
 
 import java.util.List;
+import java.util.UUID;
```
Please fix Checkstyle violations.

trustin commented on core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java:
```diff
@@ -98,7 +100,9 @@ protected AbstractRequestContextBuilder(boolean server, HttpRequest req) {
-
+        if (uuid == null) {
+            uuid = UUID.randomUUID();
+        }
```
This does not make any sense. Remove this block entirely.

trustin commented on core/src/main/java/com/linecorp/armeria/common/AbstractRequestContextBuilder.java:
```diff
@@ -433,6 +438,24 @@ protected final String path() {
+    public AbstractRequestContextBuilder uuid(UUID uuid) {
+        this.uuid = requireNonNull(uuid, "uuid");
+        return this;
+    }
```
Make sure to override this method in `{Client,Service}RequestContextBuilder` and change the return type, e.g.

```java
class ClientRequestContextBuilder ... {
    ...
    @Override
    public ClientRequestContextBuilder uuid(UUID uuid) {
        return (ClientRequestContextBuilder) super.uuid(uuid);
    }
    ...
}
```

trustin commented on core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java:
```diff
@@ -158,8 +159,8 @@ private DefaultServiceRequestContext(
-        super(meterRegistry, sessionProtocol,
-              requireNonNull(routingContext, "routingContext").method(), routingContext.path(),
+        super(meterRegistry, sessionProtocol, UUID.randomUUID(),
+              requireNonNull(routingContext).method(), routingContext.path(),
```
As I asked before already, Make sure `DefaultServiceRequestContext` constructors require `UUID` as a parameter, just like `DefaultClientRequestContext` does.

trustin commented on core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java:
```diff
@@ -184,7 +185,6 @@ private DefaultServiceRequestContext(
         // there is some delay between the actual channel read and this logging, but it's the best we can do for
         // now.
         log.requestFirstBytesTransferred();
-
```
Revert

trustin commented on core/src/main/java/com/linecorp/armeria/server/HttpServerHandler.java:
```diff
@@ -321,7 +322,6 @@ private void handleRequest(ChannelHandlerContext ctx, DecodedHttpRequest req) th
 
         final String hostname = hostname(headers);
         final VirtualHost host = config.findVirtualHost(hostname);
-
```
Revert

trustin commented on core/src/test/java/com/linecorp/armeria/common/RequestContextTest.java:
```diff
@@ -442,7 +443,7 @@ private NonWrappingRequestContext createContext(boolean addContextAwareHandler)
         DummyRequestContext() {
             super(NoopMeterRegistry.get(), SessionProtocol.HTTP,
-                  HttpMethod.GET, "/", null, HttpRequest.streaming(HttpMethod.GET, "/"), null);
+                  UUID.randomUUID(), HttpMethod.GET, "/",null, HttpRequest.streaming(HttpMethod.GET, "/"), null);
```
Missing space between `,` and `null`

trustin commented on core/src/test/java/com/linecorp/armeria/server/RouteTest.java:
```diff
@@ -23,6 +23,7 @@
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
+
```
Revert

matsumana committed:
Add Armeria version metric (#2179)

Motivation:

- Want to check Armeria version with Grafana
  FYI:
  In Prometheus best practices, this kind of metric is described as follows.
  https://prometheus.io/docs/practices/naming/
  > foobar_build_info (for a pseudo-metric that provides metadata about the running binary)
  
  e.g.
  ```
  # HELP node_exporter_build_info A metric with a constant '1' value labeled by version, revision, branch, and goversion from which node_exporter was built.
  # TYPE node_exporter_build_info gauge
  node_exporter_build_info{branch="master",goversion="go1.7.5",revision="840ba5dcc71a084a3bc63cb6063003c1f94435a6",version="0.14.0"} 1
  ```

Modifications:

- Add Armeria version metric

Result:

It's exported as follows:

Labels:
- `commit`
- `version`
- `repoStatus`

```
# HELP armeria_build_info A metric with a constant '1' value labeled by version and commit hash from which Armeria was built.
# TYPE armeria_build_info gauge
armeria_build_info{commit="74700d402178e00045b1ca23006f7f6131b2a92d",version="0.94.1-SNAPSHOT",repoStatus="clean"} 1.0
```

jyblue committed:
Add `Server` and `Date` header by default (#2188)

Motivation:

- Related #2137

Modifications:

- Modify `HttpServerHandler.handleRequest()` to add `Server` and `Date`
header by default
- Allow disabling the default behavior by adding `ServerBuilder.disableServerHeader()` and `disableDateHeader()`.
- Add `ServerConfig.isServerHeaderEnabled()` and `isDateHeaderEnabled()`
- Add test cases

Result:

- Fixes #2137

yanghun0070 committed:
Fix Checkstyle violation class
Remove UUID from DefaultServiceRequestContext

yanghun0070 committed:
Fix Checkstyle violation class
Remove UUID from DefaultServiceRequestContext

yanghun0070 committed:
Merge branch 'feature/AddUUIDFromRequest' of https://github.com/yanghun0070/armeria into feature/AddUUIDFromRequest

* 'feature/AddUUIDFromRequest' of https://github.com/yanghun0070/armeria:
  Fix Checkstyle violation class Remove UUID from DefaultServiceRequestContext
  Remove UUID from RoutingContext
  Fix UUID to this method
  Fix consistence to moving uuid right before method Fix DefaultServiceRequestContext have symmetric DefaultClientRequestContext
  Add UUID To RequestContext

# Conflicts:
#	brave/src/test/java/com/linecorp/armeria/client/brave/BraveClientIntegrationTest.java
#	core/src/main/java/com/linecorp/armeria/client/DefaultClientRequestContext.java

yanghun0070 committed:
Fix checkstyle violation

trustin commented:
Seems like the PR history has gone haywire :sob: Could you try to fix it so this pull request contains only your commits? Please consider creating a new pull request if you think it'll be easier for you.

ikhoon commented:
Close and replace with #2203
